datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/mirrormaker/KafkaMirrorMakerConnector.java

```diff
@@ -109,4 +141,142 @@ public void postDatastreamInitialize(Datastream datastream, List<Datastream> all
+        partitionDiscoveryThread.setDatastreamGroup(datastreamGroup);
+      } else {
+        //We dont need to register the thread if no partition callback is available
+        if (_partitionChangeCallback != null) {
+          partitionDiscoveryThread =
+              new PartitionDiscoveryThread(datastreamGroup);
+          partitionDiscoveryThread.start();
+          _partitionDiscoveryThreadMap.put(datastreamGroupName, partitionDiscoveryThread);
+          LOG.info("DatastreamChangeListener for {} registered", datastreamGroupName);
+        }
+      }
+    });
+  }
+
+  class PartitionDiscoveryThread extends Thread {
+    private Consumer<?, ?> _consumer;
+    private DatastreamGroup _datastreamGroup;
+    private List<String> _subscribedPartitions = new ArrayList<>();
+    private Pattern _topicPattern;
+    private boolean _initialized;
+
+
+    private PartitionDiscoveryThread(DatastreamGroup datastreamGroup) {
+      _datastreamGroup = datastreamGroup;
+      _topicPattern = Pattern.compile(
+          KafkaConnectionString.valueOf(_datastreamGroup.getDatastreams().get(0).getSource().getConnectionString()).getTopicName());
+      _initialized = false;
+    }
+
+    private void setDatastreamGroup(DatastreamGroup datastreamGroup) {
+      _datastreamGroup = datastreamGroup;
+      _topicPattern = Pattern.compile(
+          KafkaConnectionString.valueOf(_datastreamGroup.getDatastreams().get(0).getSource().getConnectionString()).getTopicName());
+    }
+
+    private List<String> getPartitionsInfo() {
+      Map<String, List<PartitionInfo>> sourceTopics = _consumer.listTopics();
+      List<TopicPartition> topicPartitions = sourceTopics.keySet().stream()
+          .filter(t1 -> _topicPattern.matcher(t1).matches()).flatMap(t2 ->
+              sourceTopics.get(t2).stream().map(partitionInfo ->
+                  new TopicPartition(partitionInfo.topic(), partitionInfo.partition()))).collect(Collectors.toList());
+
+      return topicPartitions.stream().map(TopicPartition::toString).sorted().collect(Collectors.toList());
+    }
+
+    private Consumer<?, ?> createConsumer(Properties consumerProps, String bootstrapServers, String groupId) {
+      Properties properties = new Properties();
+      properties.putAll(consumerProps);
+      properties.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+      properties.put(ConsumerConfig.GROUP_ID_CONFIG, groupId);
+      properties.putIfAbsent(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+          ByteArrayDeserializer.class.getCanonicalName());
+      properties.putIfAbsent(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+          ByteArrayDeserializer.class.getCanonicalName());
+      return _listenerConsumerFactory.createConsumer(properties);
+    }
+
+    @Override
+    public void run() {
+      Datastream datastream = _datastreamGroup.getDatastreams().get(0);
+      String bootstrapValue = String.join(KafkaConnectionString.BROKER_LIST_DELIMITER,
```

Comment:
Probably better as 
```
.stream()
.map(KafkaBrokerAddress::toString)
.collect(Collectors.joining(KafkaConnectionString.BROKER_LIST_DELIMITER));
```
instead of having the String.join wrap the Stream call chain.